drivers/usb/usbxhci/usbxhci.c

```diff
@@ -0,0 +1,572 @@
+    MaxDeviceSlots = StructParams1.MaxDeviceSlots;
+
+    DPRINT("XHCI_StartController: CapabilityRegisters - %p\n", CapabilityRegisters);
+    DPRINT("XHCI_StartController: OperRegisters - %p\n", OperRegisters);
+    DPRINT("XHCI_StartController: HC supports max %d device slot(s)\n", MaxDeviceSlots);
+
+    XhciExt->NumberOfPorts = StructParams1.MaxPorts;
+
+    /* xHCI spec says we need to perform a chip hardware reset */
+    RetStatus = XHCI_ResetController(OperRegisters);
+    if (RetStatus != MP_STATUS_SUCCESS)
+    {
+        return RetStatus;
+    }
+
+    Config.AsULONG = READ_REGISTER_ULONG(&OperRegisters->Config.AsULONG);
+
+    /* Enable all device slots */
+    Config.MaxSlotsEn = MaxDeviceSlots;
+    DPRINT("XHCI_StartController: %d device slot(s) are to be used\n", Config.MaxSlotsEn);
+
+    WRITE_REGISTER_ULONG(&OperRegisters->Config.AsULONG, Config.AsULONG);
+
+    /* TODO: There is much more to do */
+
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+XHCI_StartController(
+    _In_ PVOID XhciExtension,
+    _In_ PUSBPORT_RESOURCES Resources)
+{
+    PXHCI_HC_OPER_REGS OperRegisters;
+    PXHCI_HC_CAPABILITY_REGISTERS CapabilityRegisters;
+    XHCI_USB_COMMAND Command;
+    XHCI_USB_STATUS Status;
+    BOOLEAN IsPolling;
+    LARGE_INTEGER EndTime;
+    LARGE_INTEGER CurrentTime;
+
+    MPSTATUS RetStatus;
+    UCHAR CapLength;
+
+    DPRINT("XHCI_StartController: Starting...\n");
+
+    /* Check resources */
+    if ((Resources->ResourcesTypes & (USBPORT_RESOURCES_MEMORY | USBPORT_RESOURCES_INTERRUPT)) !=
+                                     (USBPORT_RESOURCES_MEMORY | USBPORT_RESOURCES_INTERRUPT))
+    {
+        DPRINT1("XHCI_StartController: Controller does not meet resource requirements (have %x)...\n",
+                Resources->ResourcesTypes);
+
+        return MP_STATUS_ERROR;
+    }
+
+    CapabilityRegisters = (PXHCI_HC_CAPABILITY_REGISTERS)Resources->ResourceBase;
+    CapLength = READ_REGISTER_UCHAR(&CapabilityRegisters->CapLength);
+    OperRegisters = (PXHCI_HC_OPER_REGS)((ULONG_PTR)CapabilityRegisters +
+                                                    CapLength);
+
+    Command.AsULONG = READ_REGISTER_ULONG(&OperRegisters->UsbCmd.AsULONG);
+
+    RetStatus = XHCI_InitController(OperRegisters,
+                                    Resources,
+                                    XhciExtension,
+                                    CapabilityRegisters,
+                                    CapLength);
+
+    if (RetStatus != MP_STATUS_SUCCESS)
+    {
+        return RetStatus;
+    }
+
+    Command.RunStop = 1;
+    WRITE_REGISTER_ULONG(&OperRegisters->UsbCmd.AsULONG, Command.AsULONG);
+
+    /*
+     * Just to be safe, we will wait 16ms
+     * to allow the HC to start up
+     *
+     * XXX: Is it risky not to wait?
+     */
+    KeQuerySystemTime(&EndTime);
+    EndTime.QuadPart += XHCI_POLL_TIME_SET(16);
+
+    IsPolling = FALSE;
+    while (IsPolling)
```

Comment:
With polling loops, shouldn't one also do a `YieldProcessor();` to relieve the CPU a bit?

Reply:
Question for @ThFabba : In such code, we cannot use KeDelayExecutionThread ?

Reply:
> With polling loops, shouldn't one also do a `YieldProcessor();` to relieve the CPU a bit?

Is `YieldProcessor()` a wrapper around x86's `PAUSE` instruction (on x86)? Would make sense but I could be wrong.

Reply:
> > With polling loops, shouldn't one also do a `YieldProcessor();` to relieve the CPU a bit?
> 
> Is `YieldProcessor()` a wrapper around x86's `PAUSE` instruction (on x86)? Would make sense but I could be wrong.

https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&s=YieldProcessor
https://learn.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-yieldprocessor

Reply:
I was considering leaving a similar comment. Yes, with long waits like 16ms, it probably makes sense to use either YieldProcessor or KeDelayExecutionThread. I guess it depends on the _expected_ time it would take, rather than the maximum.

Reply:
> I was considering leaving a similar comment. Yes, with long waits like 16ms, it probably makes sense to use either YieldProcessor or KeDelayExecutionThread. I guess it depends on the _expected_ time it would take, rather than the maximum.

The lines you sent is around when the controller gets started up. I did not see anything in the spec about waiting when you start up the controller but did so anyways to be safe. 16ms is probably a worse case that could occur due to, for example, host controller quirks and I don't think it should take that much time and the loop should terminate before 16ms goes by.